tutorials/application_design/c++11/publisher.cxx

```diff
@@ -0,0 +1,133 @@
+            : metrics_writer_(metrics_writer),
+              transit_writer_(transit_writer),
+              vehicle_vin_(utils::new_vin()),
+              vehicle_fuel_(100.0),
+              vehicle_route_(utils::new_route()),
+              vehicle_position_(vehicle_route_[0])
+    {
+    }
+
+    bool has_ended() const
+    {
+        return is_out_of_fuel();
+    }
+
+    void run();
+
+    friend std::string utils::to_string<>(PublisherSimulation const &sim);
+
+private:
+    dds::pub::DataWriter<VehicleMetrics> metrics_writer_;
+    dds::pub::DataWriter<VehicleTransit> transit_writer_;
+
+    std::string vehicle_vin_;
+    double vehicle_fuel_;
+    utils::CoordSequence vehicle_route_;
+    Coord vehicle_position_;
+
+    bool is_out_of_fuel() const
+    {
+        return vehicle_fuel_ <= 0.0;
+    }
+
+    bool is_on_standby() const
+    {
+        return vehicle_route_.empty();
+    }
+};
+
+void PublisherSimulation::run()
+{
+    while (!has_ended()) {
+        metrics_writer_.write(VehicleMetrics { vehicle_vin_, vehicle_fuel_ });
+
+        transit_writer_.write(VehicleTransit { vehicle_vin_,
+                                               vehicle_position_,
+                                               vehicle_route_ });
+
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+
+        if (is_on_standby()) {
+            std::cout << "Vehicle '" << vehicle_vin_
+                      << "' has reached its destination, now moving to a "
+                         "new location..."
+                      << std::endl;
+            vehicle_route_ = utils::new_route();
+            vehicle_route_[0] = vehicle_position_;
+        }
+
+        vehicle_fuel_ -= 10 * utils::random_stduniform();
+        vehicle_position_ = vehicle_route_.front();
+        vehicle_route_.erase(vehicle_route_.begin());
+
+        if (is_out_of_fuel()) {
+            vehicle_fuel_ = 0.0;
+            std::cout << "Vehicle '" << vehicle_vin_ << "' ran out of fuel!"
+                      << std::endl;
+        }
+    }
+}
+
+template<>
+std::string utils::to_string(PublisherSimulation const &sim)
+{
+    std::ostringstream ss;
+    ss << "PublisherSimulation(vehicle_vin: " << sim.vehicle_vin_;
+    ss << ", vehicle_fuel: " << sim.vehicle_fuel_;
+    ss << ", vehicle_route: " << to_string(sim.vehicle_route_);
+    ss << ", vehicle_position: " << to_string(sim.vehicle_position_) << ")";
+    return ss.str();
+}
+
+int main(int argc, char **argv)
+{
+    rti::domain::register_type<VehicleMetrics>();
+    rti::domain::register_type<VehicleTransit>();
+
+    dds::core::QosProvider qos_provider("../VehicleModeling.xml");
+
+    dds::domain::DomainParticipant participant =
+            qos_provider.extensions().create_participant_from_config(
+                    "ParticipantLibrary::PublisherApp");
+
+    using MetricsWriter = dds::pub::DataWriter<VehicleMetrics>;
+    MetricsWriter metrics_writer(
```

Comment:
I kept error-handling to a minimum because I wanted the documentation and code to be verbatim 1:1. I'm not checking for `None` or exceptions in Python either.

I can make this change, but given that this is a tailored example with a fixed name in XML, I also think we can leave it be as-is.

If you think I'm declining too much feedback on better practises, perhaps we could add a disclaimer in the C++-specific `README.md`.

Reply:
I think it's this case we have to check for null. In C you can just say "we're not checking for any errors" but here it's very misleading to assume that errors are reporter as exceptions and not handle the few conditions that are not.

You just need to do this
```
if (metris_writer == dds::core::null) {
    throw std::runtime_error("metrics_writer not found");
}
```